builtin/authenticator/oidcropc/Authenticator.go

```diff
@@ -0,0 +1,104 @@
+package oidcropc
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/viper"
+)
+
+// Authenticator struct represents OIDC options for SMK Authentication.
+type Authenticator struct {
+	OIDCTokenEndpoint      string
+	OIDCClientID           string
+	OIDCClientSecret       string
+	OIDCAlternatePrincipal bool
+}
+
+// Init method is used to ingest config of Authenticator
+func (a *Authenticator) Init(config *viper.Viper) error {
+	neededEntries := []string{
+		"oidcTokenEndpoint",
+		"oidcClientID",
+		"oidcClientSecret",
+	}
+
+	var missingEntriesLst []string
+	for _, entry := range neededEntries {
+		if !config.IsSet(entry) {
+			missingEntriesLst = append(missingEntriesLst, entry)
+		}
+	}
+	if len(missingEntriesLst) > 0 {
+		missingEntries := strings.Join(missingEntriesLst, ", ")
+		return fmt.Errorf("Missing config entries (%s) for Authenticator", missingEntries)
+	}
+
+	a.OIDCTokenEndpoint = config.GetString("oidcTokenEndpoint")
+	a.OIDCClientID = config.GetString("oidcClientID")
+	a.OIDCClientSecret = config.GetString("oidcClientSecret")
+	a.OIDCAlternatePrincipal = config.GetBool("oidcAlternatePrincipal")
+
+	return nil
+}
+
+type oidcToken struct {
+	Token string `json:"access_token"`
+}
+
+// Login method is used to check if a couple of user/password is valid in LDAP.
+func (a *Authenticator) Login(user, password string) (valid bool, swapuser string, err error) {
+
+	payload := strings.NewReader("grant_type=password&" +
```

Comment:
Using https://golang.org/pkg/net/url/#Values is more idiomatic

Reply:
Done